ledger/metrics/src/main/scala/com/daml/metrics/MetricHandle.scala

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.metrics
+
+import java.util.concurrent.TimeUnit
+import cats.data.EitherT
+import com.codahale.metrics.MetricRegistry.MetricSupplier
+import com.codahale.metrics.Snapshot
+import com.codahale.{metrics => codahale}
+
+import scala.concurrent.{Future, blocking}
+
+sealed trait MetricHandle[T <: codahale.Metric] {
+  def name: String
+  def metric: T
+  def metricType: String // type string used for documentation purposes
+}
+
+object MetricHandle {
+
+  trait Factory {
+
+    def prefix: MetricName
+
+    def registry: codahale.MetricRegistry
+
+    def timer(name: MetricName): Timer = Timer(name, registry.timer(name))
+
+    def varGauge[T](name: MetricName, initial: T): VarGauge[T] =
+      addGauge(name, Gauges.VarGauge[T](initial), _.updateValue(initial))
+
+    @SuppressWarnings(Array("org.wartremover.warts.AsInstanceOf"))
+    private def addGauge[T <: codahale.Gauge[M], M](
+        name: MetricName,
+        newGauge: => T,
+        resetExisting: (T => Unit),
+    ): Gauge[T, M] = blocking {
+      synchronized {
+        val res: Gauge[T, M] = Option(registry.getGauges.get(name: String)) match {
+          case Some(existingGauge) => Gauge(name, existingGauge.asInstanceOf[T])
+          case None =>
+            val gauge = newGauge
+            // This is not idempotent, therefore we need to query first.
+            registry.register(name, gauge)
+            Gauge(name, gauge)
+        }
+        resetExisting(res.metric)
+        res
+      }
+    }
+
+    def gauge[T](
+        name: MetricName,
+        gaugeSupplier: MetricSupplier[codahale.Gauge[_]],
+    ): Gauge[codahale.Gauge[T], T] =
+      registry.synchronized {
```

Comment:
I am wondering why it's not symmetrical to `addGauge` synchronisation? :) probably question to original author of the code

Reply:
also while the `addGauge` is marked `blocking` but this one is not

Reply:
Actually the former (`addGauge`) is from `canton` while the latter (`gauge`) is from `daml-sdk` (was previously as [registerGauge](https://github.com/digital-asset/daml/blob/10f874f465cb90876f53fd539b4a9f196b071e8a/ledger/metrics/src/main/scala/com/daml/metrics/package.scala#L11-L20))

Reply:
What about adding `blocking` to the latter as well and add a `registry.synchronized` to the former (to make it more fine-grained and lock the registry and not the class instance)?

Reply:
Moreover, should I rename `gauge` to `gaugeWithSupplier`?

Reply:
I think by the arguments I can see that both methods are identical, just providing slightly different API. Can you just leave a single method?

Reply:
Basically would be good to generalise both methods and expose just one or 2 but going to the same generalised method

Reply:
What about 3f0be6a1505d3028b601f8add14866e07b634607?